### DISCUSSION TITLE
How to import script from `node_modules`

### DISCUSSION THREAD

**Discussion by blue32a** [vercel/next.js#45684] [Help] [answered]
HI!

I installed Bootstrap with Next.js **13.1.6**. `npm install bootstrap`
I was able to import the css.

[Import styles from node_modules](https://nextjs.org/docs/basic-features/built-in-css-support#import-styles-from-node_modules)

```
// pages/_app.js
import 'bootstrap/dist/css/bootstrap.css'

export default function MyApp({ Component, pageProps }) {
  return <Component {...pageProps} />
}
```

How do I import `bootstrap.min.js` from `node_modules`?
Or should I stop `npm install bootstrap` and use CDN?

```
import Head from 'next/head'
import Script from 'next/script'

export default function App({ Component, pageProps }: AppProps) {
  return (
    <>
      <Head>
        <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/x.y.z/css/bootstrap.min.css" />
      </Head>
      <Script src="https://cdn.jsdelivr.net/npm/bootstrap@x.y.z/dist/js/bootstrap.min.js" />
      <Component {...pageProps} />
    </>
  )
}
```

I look forward to hearing from you.

**Comment by icyJoseph** [comment]
From the Bootstrap docs:

```js
// Import all of Bootstrap's JS
import * as bootstrap from 'bootstrap'
```

And I believe you can import it where ever you need it, but you need to install:

```
 npm i --save-dev --save-exact @popperjs/core
```

I'd recommend reading https://getbootstrap.com/docs/5.3/getting-started/webpack/

**Reply by blue32a** [reply]
Thank you, I found the answer in the issue.

**Comment by blue32a** [comment, marked as answer]
I found the answer.

> You should never use jquery and bootstrap's JavaScript in a React application. This is true for all React apps—not only Next.js.
> 
> Using a library like [React Bootstrap](https://react-bootstrap.github.io/) gives you minimal bundle size and full React support.
> 
> Mixing Bootstrap JS (+ jQuery) will always result in application bugs because you're mutating elements that React controls.
> 
> If you must use Bootstrap's JS and jQuery, note that these are client-side only libraries and must be loaded as so:
> 
> import "bootstrap/dist/css/bootstrap.css";
> import { useEffect } from "react";
> 
> export default function MyApp({ Component, pageProps }) {
>   useEffect(() => {
>     import("jquery").then($ => {
>       // jQuery must be installed to the `window`:
>       window.$ = window.jQuery = $;
>       return import("bootstrap");
>     });
>   }, []);
> 
>   return <Component {...pageProps} />;
> }
> I highly advise you do not use this in your application. Use [React Bootstrap](https://react-bootstrap.github.io/) instead!

https://github.com/vercel/next.js/issues/10627#issuecomment-589686848

**Reply by icyJoseph** [reply]
Well, frontend development is such a massive force that it made that comment a **tiny bit** outdated.

It is true that it is better to learn the React way, over just trying to re-iterate over what's other frameworks have been asking from you.

> When learning a new technology, it’s natural to try and squeeze it into a familiar shape.

However, Bootstrap 5 dropped jQuery and ships their own Bootstrap framework now. 

This means that you can treat Bootstrap as an external API, and integrate with it via refs, naturally, not that you couldn't before, but now it is made for that.

On top of that there has always been `react-bootstrap` and `reactstrap` which sort of glued bootstrap's markdown and CSS under the React way of working.

I'd still take any of those React Bootstrap frameworks over directly working with the new Bootstrap API, but it is there, and no longer requires jQuery.

**Reply by icyJoseph** [reply]
So yeah, don't import it, unless you really want/need to 😅

**Reply by blue32a** [reply]
The package we wanted to use required `Bootstrap v4`. We will consider using React Bootstrap frameworks. Thank you.
